std-bits/table/src/main/java/org/enso/table/parsing/NumberParser.java

```diff
@@ -0,0 +1,279 @@
+        for (int i = 0; i < sourceStorage.size(); i++) {
+            var text = sourceStorage.getItemBoxed(i);
+            if (text == null) {
+                builder.appendNulls(1);
+            } else {
+                var value = innerParseSingleValue(text, pattern);
+                if (value != null) {
+                    builder.appendNoGrow(value);
+                } else {
+                    if (aggregator == null) {
+                        return i;
+                    }
+
+                    aggregator.reportInvalidFormat(text);
+                    builder.appendNulls(1);
+                }
+            }
+        }
+        return -1;
+    }
+
+    @Override
+    protected Builder makeBuilderWithCapacity(int capacity) {
+        return allowDecimal
+                ? NumericBuilder.createDoubleBuilder(capacity)
+                : NumericBuilder.createLongBuilder(capacity);
+    }
+
+    private Object innerParseSingleValue(String text, Pattern pattern) {
+        if (allowDecimal) {
+            var trimmed = trimValues ? text.trim() : text;
+            if (trimmed.equals("NaN")) {
+                return Double.NaN;
+            }
+            if (trimmed.equals("Infinity")) {
+                return Double.POSITIVE_INFINITY;
+            }
+            if (trimmed.equals("-Infinity")) {
+                return Double.NEGATIVE_INFINITY;
+            }
+        }
+
+        var parsed = pattern.matcher(text);
+        if (!parsed.matches()) {
+            return null;
+        }
+
+        try {
+            var sign = parsed.group("sign");
+            var sign_value = sign != null && !sign.equals("+") ? -1 : 1;
+
+            var integer = parsed.group("integer").replaceAll("\\D", "");
+
+            if (!allowLeadingZeros && integer.length() > 1 && integer.charAt(0) == '0') {
+                return null;
+            }
+
+            if (allowDecimal) {
+                var decimal = parsed.group("decimal");
+                var decimalPrepared = decimal == null ? "" : ("." + decimal.substring(1));
+
+                if (integer.equals("") && decimalPrepared.equals("")) {
+                    return null;
+                }
+
+                integer = integer.equals("") ? "0" : integer;
+
+                if (allowScientific) {
+                    var exp = parsed.group("exp");
+                    if (exp != null) {
+                        if (integer.length() > 1) {
+                            return null;
+                        }
+                        decimalPrepared = decimalPrepared + exp;
+                    }
+                }
+
+                return sign_value * Double.parseDouble(integer + decimalPrepared);
+            }
+
+            return integer.equals("") ? null : sign_value * Long.parseLong(integer);
+        } catch (NumberFormatException e) {
+            return null;
+        }
```

Comment:
How likely are we to hit this exception?

I remember the trouble we had with parsing dates throwing too many exceptions under the hood, so I'm cautious.

I assume there are 'preflight checks' that should ensure that in practice only valid patterns get to the point of `parseDouble`/`parseLong` call, is that correct?

Reply:
I think exceptionally unlikely.

The parts must already have conformed to a valid format before it is even called.

Reply:
Ok, perfect. That's what I thought but I wanted to make sure.

I have a very optional suggestion: since we care about performance, we probably want to detect if some regression were to enable this `catch` branch.

To avoid this I'd propose one of:
1. change `return null` to `throw IllegalStateException` so that it explodes and we see that something is wrong,
2. add a warning using SLF4J logger stating that this is unexpected.

I like (1) because if we are confident of its correctness, it will only fire if there are serious bugs likely breaking the numeric parsing anyway so we probably want to detect it ASAP. But I appreciate we may want to try not breaking the flow of program. Such a bug will result in worst case failing to parse some numbers - probably not the worst kind of data corruption, so I think it's OK to avoid failing if you think that's better - then maybe (2) could provide us at least some immediate feedback in the logs if we were ever to encounter this.

What do you think?